readthedocs/templates/search/elastic_search.html

```diff
@@ -25,218 +27,228 @@
+                      {{ hit|key:'_source'|key:'project' }} - {{ hit|key:'_source'|key:'title' }}
+                    </a>
+
+                    {% for inner_hit in hit.inner_hits %}
+                      {% if inner_hit.type == 'domains' %}
+                        <p>
+                          <a href="{% doc_url hit|key:'_source'|key:'project'|get_project hit|key:'_source'|key:'version' inner_hit.source.doc_name %}?highlight={{ query }}#{{ inner_hit.source.anchor }}">
+                            {% if inner_hit.source.display_name|length >= 2 %}
+                              ({{ inner_hit.source.role_name }}) {{ inner_hit.source.display_name}}
+                            {% else %}
+                              {{ inner_hit.source.role_name }}
+                            {% endif %}
                           </a>
-                          {% for fragment in result.meta.highlight.description %}
-                            <p class="fragment">
-                            ...{{ fragment|safe }}...
-                            </p>
-                          {% empty %}
-                          <p><em>No description for project.</em></p>
-                          {% endfor %}
-
-                      {% elif 'domain' in result.meta.index %}
-
-                        <a href="{% doc_url result.project|get_project result.version result.doc_name %}?highlight={{ query }}#{{ result.anchor }}">
-                          {{ result.project }} - {% if result.meta.highlight.name|length %} {{ result.meta.highlight.name.0|safe }} {% else %} {{ result.name }} {% endif %}
-                        </a>
+                        </p>
                         <p class="fragment">
-                          {% if result.type_display %}
-                            {{ result.type_display|capfirst }}
+                          {% if inner_hit.highlight|get_key_or_none:"domains.name" %}
+                            {% with l=inner_hit.highlight|get_key_or_none:"domains.name" %}
+                              {{ inner_hit.source.type_display }} -- {{ l.0 }} -- {{ inner_hit.source.doc_display }}
+                            {% endwith %}
                           {% else %}
-                            {{ result.role_name }}
+                            {{ inner_hit.source.type_display }} -- {{ inner_hit.source.name }} -- {{ inner_hit.source.doc_display }}
                           {% endif %}
-                          in
-                          {% if result.doc_display %}
-                            {{ result.doc_display|capfirst }}
+                        </p>
+
+                      {% elif inner_hit.type == 'sections' %}
+                        <p>
+                          <a href="{% doc_url hit|key:'_source'|key:'project'|get_project hit|key:'_source'|key:'version' hit|key:'_source'|key:'full_path' %}?highlight={{ query }}#{{ inner_hit.source.id }}">
```

Comment:
Things are getting complex (too many `ifs` and `elses`).
I suggest to remove `_source` and `highlight` field and preprocess the values from server side.
Like: if a result is something like this --
```json
{
    "_source": {
        "x": "a",
        "y": "b"
    },
    "highlight": {
        "x": "<em>a</em>",
    }
}
```

Then in the client side we have too add an `if` and test if the highlighted value of `x` and `y` are present -- if present, use that. And if absent, use the values from `_source`. This complicates things.
If we preprocess this on server side and send the result like this --

```json
{
    "x": "<em>a</em>",
    "y": "b"
}
```

Then it would be much simpler.
Also, this would also resolve the problem you stated in this comment.
What are your thoughts on this? @ericholscher

Reply:
This is also resolved now. :smile: